aten/src/ATen/Quantizer.h

```diff
@@ -0,0 +1,182 @@
+ */
+struct C10_API NonUniformQuantizer : public Quantizer {
+  virtual ~NonUniformQuantizer() {}
+
+};
+
+// There is also StochasticQuantizer which is uniform but not affine
+
+/**
+ * AffineQuantizer uses affine transformation to do quantization.
+ *
+ * For quantize:
+ * Y = clamp((X * scale + zero_point, min, max)
+ * For dequantize:
+ * X = (Y - zero_point) / scale
+ */
+struct C10_API AffineQuantizer : public UniformQuantizer {
+  virtual ~AffineQuantizer() {}
+
+};
+
+/**
+ * SymmetricQuantizer is similar to AffineQuantizer except that it
+ * does not have zero_point
+ *
+ * For quantize:
+ * Y = clamp(X * scale, min, max)
+ * For dequantize:
+ * X = Y / scale
+ */
+struct C10_API SymmetricQuantizer : public UniformQuantizer {
+  virtual ~SymmetricQuantizer() {}
+
+};
+
+/**
+ * PerLayerSymmetricQuantizer stores a single scale number which is
+ * used for quantizing all the values in the given Tensor
+ */
+struct C10_API PerLayerSymmetricQuantizer: public SymmetricQuantizer {
+  PerLayerSymmetricQuantizer() {}
+  PerLayerSymmetricQuantizer(float scale) : scale_(scale) {}
+  virtual ~PerLayerSymmetricQuantizer() {}
+  float scale_{1.0};
+};
+
+/**
+ * PerChannelSymmetricQuantizer stores a vector of scale number and
+ * applys symmetric quantization using different scales.
+ */
+struct C10_API PerChannelSymmetricQuantizer: public SymmetricQuantizer {
+  PerChannelSymmetricQuantizer() {}
+  virtual ~PerChannelSymmetricQuantizer() {}
+
+  std::vector<float> scales_;
+  int64_t channel_axis_;
+};
+
+/**
+ * PerLayerAffineQuantizer stores a scale and a zero_point, which is used for
+ * all the values in the Tensor.
+ */
+struct C10_API PerLayerAffineQuantizer: public AffineQuantizer{
+  PerLayerAffineQuantizer(float scale, uint8_t zero_point): scale_(scale), zero_point_(zero_point) {}
+
+  virtual QTensor quantize(RealTensor tensor);
+  virtual RealTensor dequantize(QTensor tensor);
+
+  float scale() {
+    return scale_;
+  }
+
+  int32_t zero_point() {
+    return zero_point_;
+  }
+
+  float scale_{1.0};
+  uint8_t zero_point_{0};
+};
+
+/**
+ * PerChannelAffineQuantizer is the same as PerLayerAffineQuantizer
+ * except that we have an independent scale and zero_point parameter
+ * for each channel.
+ */
+struct C10_API PerChannelAffineQuantizer: public AffineQuantizer {
+  PerChannelAffineQuantizer() {}
+
+  std::vector<float> scales_;
+  std::vector<uint8_t> zero_points_;
+  int64_t channel_axis_;
+};
+
+// This is an internal utility function for getting at the QTensorImpl,
+// You should only use this for writing low level
+// setters/getters for QTensorImpl fields; otherwise, you should use
+// the low level setters/getters that were implemented using this.
+// This may be called repeatedly, so make sure it's pretty cheap.
+QTensorImpl* get_qtensorimpl(const QTensor& self);
+
+/* Some Helper Functions */
+template <class T>
+inline T Round(const T x) {
+  return std::nearbyint(x);
```

Comment:
why do you need a wrapper?

Reply:
this is copied from other pr, but I'm deleting ChooseParams, so this one will be deleted as well